receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,512 @@
+
+		authProtocol := getAuthProtocol(cfg.AuthType)
+		securityParams.AuthenticationProtocol = authProtocol
+		securityParams.AuthenticationPassphrase = cfg.AuthPassword
+
+		privProtocol := getPrivacyProtocol(cfg.PrivacyType)
+		securityParams.PrivacyProtocol = privProtocol
+		securityParams.PrivacyPassphrase = cfg.PrivacyPassword
+	}
+	client.SetSecurityParameters(securityParams)
+}
+
+// getAuthProtocol gets gosnmp auth protocol based on config auth type
+func getAuthProtocol(authType string) gosnmp.SnmpV3AuthProtocol {
+	var authProtocol gosnmp.SnmpV3AuthProtocol
+
+	switch strings.ToUpper(authType) {
+	case "MD5":
+		authProtocol = gosnmp.MD5
+	case "SHA":
+		authProtocol = gosnmp.SHA
+	case "SHA224":
+		authProtocol = gosnmp.SHA224
+	case "SHA256":
+		authProtocol = gosnmp.SHA256
+	case "SHA384":
+		authProtocol = gosnmp.SHA384
+	case "SHA512":
+		authProtocol = gosnmp.SHA512
+	}
+
+	return authProtocol
+}
+
+// getPrivacyProtocol gets gosnmp privacy protocol based on config privacy type
+func getPrivacyProtocol(privacyType string) gosnmp.SnmpV3PrivProtocol {
+	var privacyProtocol gosnmp.SnmpV3PrivProtocol
+
+	switch strings.ToUpper(privacyType) {
+	case "DES":
+		privacyProtocol = gosnmp.DES
+	case "AES":
+		privacyProtocol = gosnmp.AES
+	case "AES192":
+		privacyProtocol = gosnmp.AES192
+	case "AES192C":
+		privacyProtocol = gosnmp.AES192C
+	case "AES256":
+		privacyProtocol = gosnmp.AES256
+	case "AES256C":
+		privacyProtocol = gosnmp.AES256C
+	}
+
+	return privacyProtocol
+}
+
+// Connect uses the goSNMP client's connect
+func (c *snmpClient) Connect() error {
+	return c.client.Connect()
+}
+
+// Close uses the goSNMP client's close
+func (c *snmpClient) Close() error {
+	return c.client.Close()
+}
+
+// GetScalarData retrieves scalar metrics from passed in scalar OIDs. The returned data
+// is then also passed into the provided function.
+// Note: These OIDs must all end in ".0" for the SNMP GET to work correctly
+func (c *snmpClient) GetScalarData(oids []string, processFn processFunc, scraperErrors *scrapererror.ScrapeErrors) {
+	// Nothing to do if there are no OIDs
+	if len(oids) == 0 {
+		return
+	}
+
+	// Group OIDs into chunks based on the max amount allowed in a single SNMP GET
+	chunkedOIDs := chunkArray(oids, c.client.GetMaxOids())
+
+	// For each group of OIDs
+	for _, oidChunk := range chunkedOIDs {
+		// Note: Not implementing GetBulk as I don't think it would work correctly for the current design
+		packets, err := c.client.Get(oidChunk)
+		if err != nil {
+			scraperErrors.AddPartial(len(oidChunk), fmt.Errorf("problem with getting scalar data: problem with SNMP GET for OIDs '%v': %w", oidChunk, err))
+			// Prevent getting stuck in a failure where we can't recover
+			if strings.Contains(err.Error(), "request timeout (after ") {
+				if err = c.Close(); err != nil {
+					c.logger.Warn("Problem with closing connection while trying to reset it", zap.Error(err))
+				}
+				if err = c.Connect(); err != nil {
```

Comment:
This connection strategy seems odd to me. We're using a default timeout equivalent to the poll interval, but it looks like we also don't even try to connect until we time out? Won't that result in long delays on a regular basis? Maybe I'm missing something.

Reply:
@djaglowski so this logic is for if a connection has a problem during a collection. without reconnecting, the snmp library connection stays borked and will continue to timeout for all future snmp calls. (i noticed this when toggling my vpn on/off).

now i am currently planning on having the scraper call the client's "Connect()" on start and "Close()" on shutdown. maybe this is dumb and i should instead just connect/close on every single use of the client get methods. this would improve the scenario where a damaged connection happens in between collections (and we wouldn't have to wait for the first timeout to reconnect). i'll plan on switching to this

Reply:
@djaglowski eh, i'm actually on the fence for connecting every time for very short collection intervals. what are your thoughts?

Reply:
How about connecting at the beginning of every poll interval, and then again if there's a problem?

Also, I think it is probably worth having a much shorter timeout. With the current timeout being as long as the poll interval, this effectively means that any problem with the connection will result in no additional data collected until the next interval. Even then, does it return the data it had collected before the problem? Seems like we can set a tighter expectation.

Reply:
@djaglowski i'll make some changes to this. for the record, even though you're right timeouts could spill over into the future collection intervals, we would still report anything before the timeout when everything eventually finishes

Reply:
@djaglowski ok, actually not much needs to change here with your suggestion. the scraper has access to `Connect()` and `Close()` and can reconnect on every scrape. I'll keep the logic while in the middle of grabbing SNMP metrics to reset if we notice a connection problem. 

so all i really did was to change the timeout to 5s

Reply:
Ok, sounds good. It probably makes sense to make the timeout configurable, but that can be in a future PR.